### ISSUE TITLE
Preserve comments across newlines?

### ISSUE THREAD

**Issue by cprice404** [typesafehub/config#149]
Hi,

I've been hacking a bit on porting this library to Ruby ( https://github.com/cprice404/ruby-typesafe-config/blob/master/README.md ).  One of the features I'm interested in is the ability to load a config file, modify the data programatically, and write the file back out to disk.  It's fine if the formatting changes a bit during the round trip, but retaining as many comments as possible is a huge win and a big part of why I am a fan of the Java library.

However, I'm a bit confused as to the expected behavior for comments that are separated by a newline.  Here's an example config file:

```
# These are some opening comments

# These are some additional opening comments
foo.bar {
  baz = 42
}
```

Now, if I load this via ConfigFactory.parseFile, and then render it back to a string (with render options +formatted, +comments, -json, -origincomments), I get this:

```
foo {
    # These are some additional opening comments
    bar {
        baz=42
    }
}
```

I understand why the comment got moved from the top of the file and is instead attached directly to "bar", but, I was a bit surprised that the very first comment line that had been at the beginning of the file just disappeared entirely.  Reading some of the comments in the source code, it sounded like there was some expectation that these comments would be grouped together despite the newline and would all be attached to "bar".

Is the current behavior expected?  Would it be possible to make it save comments across newlines?

Failing that, might it be worth considering any comments that appear at the very beginning of the file (followed by a blank line) to be "header" comments that would always be rendered at the top of the output during a render operation?

**Comment by cprice404** [comment]
@havocp do you mind if I resuscitate this thread?  We are about to ramp up our usage of (and, accordingly, our development and maintenance of) the ruby port.  The handful of things that we need that the library doesn't already do are mostly around round-tripping the config files back to disk.  In order of our priority, the main things are:

1. Preserving the order that sections and values are read in from the file
2. Potentially maintaining some basic syntactical things; e.g., if we used a colon delimiter between a key/value pair in a map, it'd be nice if the colon would survive the roundtrip (don't care about whitespace formatting)
3. This newline comment stuff that I pinged you about last year.

We will have some developer resources that we can allocate to this work.  In my ideal world, we'd start by getting PRs accepted into your Java code base, and then we'd backport them to the Ruby codebase.  It's still a goal of mine to try to keep the code in a state where it's as much of a line-by-line port as possible so that we will have the opportunity to port new changes from the Java library into the Ruby one easily, but if for whatever reason we aren't able to come up with a mutually agreeable solution for implementation details, we can always fall back to doing something directly against the Ruby codebase.

So... I guess my questions are: are you open to discussion on those (and possibly a few other similar) features?  If so, what's the best way to proceed?  I could open new github issues for each, brainstorm on implementation ideas with you until we came up with something that you were happy with, and then work on getting PRs up for review?

**Comment by havocp** [comment]
I'm open to discussion, but I guess the caution I'd give is that I think this is a lot of work. I think it would take me a week or two, and it will probably take someone else a bit longer and include a few iterations in response to discussion and review comments. (Draft PR, comments, second draft PR, comments, finish it up, etc.) The library currently has pretty complete docs and test coverage too, and I'd want to keep that, so the time budget has to include it.

If that doesn't scare you off, some implementation thoughts. I'll start with some related issues where there's been some discussion, and several of these could maybe be addressed with some of the same changes.

On #225 I floated the idea of an "AST API" (abstract syntax tree) which I think could be the same thing as the "formatted template data structure" I mentioned earlier on this issue - and thus maybe AST is the wrong term, but I don't know. Whatever we call it, this "AST" would contain all the formatting details, but would not have friendly convenience functionality for the `Config` use case, and would not be able to merge (no `withFallback`) or resolve (no `resolve`). We're representing a text file not a config tree. This AST or "template" would be for the "I want the raw syntax" use case.

`ConfigOrigin` on a `ConfigValue` could (at least optionally) simply point back into the source AST. To implement modify in load/modify/save, we'd follow the value's `ConfigOrigin` back and replace the AST nodes the value came from.

#193 and #169 both want public ways to create and modify `ConfigOrigin`.

#178 is the "roundtrip" (load-edit-save) issue.

A complication is that a given `Config` (or more precisely, `ConfigValue`) could come from N of these AST documents, of course. I think that's manageable, just have to pick rules for how editing operations behave ("pick source doc of the value that 'won' in the merge process" maybe).

Implementation-wise, my prediction is that the AST API needs to replace or underlie, rather than extend, the current (private, ugly) token stuff: https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Tokens.java

Performance considerations: I'm not sure what they will be. I don't know whether ConfigOrigin that's simply a pointer into an AST would end up being more or less memory overhead than ConfigOrigin as it exists today, or about the same. I can imagine it being less. In that case we could dump the existing SimpleConfigOrigin.

I'm not sure exactly what our "AST" or "formatted template" would look like. Possibly it's just a giant string, with "boundaries" tracked? i.e. we have this big string, and then we have a data structure that locates all tokens and line breaks and so on within it, so each "boundary" is a type tag for what's at the boundary, plus the index into the big string? In that world it could even be smaller and more efficient than the token stuff, who knows. Tokens could simply be pointers into the big string, and ConfigOrigin simply pointers to tokens ... so instead of "tokenizing" as we do now by throwing away lots of information, we could "tokenize" by recording all the boundaries in the source text, but keeping the entire source text.

That's just an idea though. I can't promise that any of these ideas will work out well, without trying them myself...

How to proceed - 

 * I'd dig in to the code and sketch your own ideas and get a sense of what you (or your chosen developers) think would work
 * maybe write up some notes similar to mine above, maybe more comprehensible, on how you're thinking of proceeding, we can discuss that and also try to figure out some step-by-step (so it isn't all one huge patch, if we can figure out how to break it down)
 * try actually proceeding on the first step and submit a PR

or something ... I can't predict how it will go :-)

I've made this sound really scary I guess. I feel like maybe it isn't in reality once our heads are wrapped around it - we could call the project "changing tokens from their current information-discarding representation to a conceptual chopping-up of the input string, so ConfigOrigin can have full knowledge of the input"  :-) ? I don't know, that's only one approach though.

One way to get started without changing every bit of code could be:

 * create the new "boundaries in a string" AST/template data structure
 * convert impl/Tokenizer.java to mark boundaries instead of generating a token stream
 * write a thing to convert the AST/template thing to a token stream
 * then the existing Parser.java doesn't have to change, nor do the unit tests that use the token API directly, but we have the new AST/template thing in the pipeline
 * could review/merge a PR here - with no new public API, and no changes to existing tests, and no changes to implementation outside of the tokenizer
 * now connect ConfigOrigin back to the AST/template thing and work out public API to do the desired stuff ... simple first features could be "save the file again unmodified" and then "change one ConfigValue and also swap out its AST nodes, then save"
 * could review/merge a PR or PRs here (one per feature)
 * then finally, purely as an optimization, consider making Parser.java work more directly with the AST/template stuff and potentially improve/eliminate/reduce the token stream layer

Going about it that way could keep the changes very localized; you're inserting a phase in the parsing pipeline between reading the byte stream and generating a token stream.

But again, this could all be nonsense. Basically somebody just has to try some things and be willing to try again if it doesn't work out. It could be wise to think of at least a couple ideas besides mine before falling in love with any.

**Comment by cprice404** [comment]
@havocp thanks for the thoughtful reply!

> I'm open to discussion, but I guess the caution I'd give is that I think this is a lot of work. I think it would take me a week or two, and it will probably take someone else a bit longer and include a few iterations in response to discussion and review comments. (Draft PR, comments, second draft PR, comments, finish it up, etc.) The library currently has pretty complete docs and test coverage too, and I'd want to keep that, so the time budget has to include it.

Totally agree with all of the above.  Wouldn't really want to approach it any other way, and definitely wouldn't expect you to accept PRs that eschewed those considerations :)

Thanks for your insights on the possible implementation paths.  They're very helpful.  We'd originally been thinking about some more crude approaches, e.g. just adding an "order" integer to the ConfigValue or ConfigOrigin classes and then sorting on that during a write.  This wouldn't preserve all formatting, but it'd potentially solve the ordering problem, which is currently our main concern.

For things like preserving ":" or "=" I was (without giving it the amount of thought that it really deserves) thinking of maybe just having a setter called "setMapKeyValueDelimiter" or something, which would accept an enum for COLON, EQUALS, NONE, and would default to whatever the library currently uses when writing out a file (which I can't recall because it's been a while since I had my hands in this code).

Then, for the comment stuff, I'd just been thinking of some heuristic stuff like what you and I discussed earlier in this thread.

I recognize that all of those are fairly crude hacks compared to the more complete solution that you're getting at, and your solution would be better all around.

So... we should have some non-trivial developer resources to assign to take a swing at this.  Not infinite, but something more than just me hacking on the weekend :)  I will try to carve out some time next week to sit down with our developer who would be looking into this stuff, and we'll re-familiarize ourselves with the code and with the relationship between the Tokenizer and the ConfigOrigin.  Assuming we can grok that and then more fully comprehend the path you're proposing, we'll try to put together a game plan for the first few bullet points that you described, and check back in with you.  If we're all on the same page then we can start working towards an initial PR.

I'm sure we'll probably have lots of dumb questions along the way, so I'll ping you :)

Thanks again for taking the time to discuss!

**Comment by havocp** [comment]
Sounds good - dig in on it a bit with your developer maybe and see what you guys think. I do have some sympathy for the piecemeal quick hacks and heuristics approach, but I feel like it might be a little neverending; we'll be punting corner cases, and naturally people will file issues wanting to fix those, and we still won't be preserving lots of formatting, and naturally people will file issues wanting to preserve more stuff, and so it will tend to be a bottomless hole of special cases.  And we do know that people would like to load/edit/save and preserve *all* the formatting, ideally. So if we can wrap our heads around nuking the whole category of problems like this once and for all... well, it has some appeal. But, we have to be able to grok it and see our way to how the complete solution would work.

If we can break it down into localized, reviewable steps - like building this "boundaries" data structure but then generating the existing tokens from that, so the parser and unit tests don't change in an initial PR - maybe it doesn't look too intimidating. An initial goal could be to have this data structure where you can load/save the file unchanged (a simple thing to unit test), and then next goal is to be able to replace one token and resave (using the internal data structure directly), and then we work our way back up to the public API...

**Comment by cprice404** [comment]
@havocp just as a quick update--still planning on devoting some time to this, just haven't gotten a chance to sink our teeth in yet due to some other urgent issues that have popped up.

**Comment by havocp** [comment]
Thanks, no problem. We're experimenting with a gitter chat room: https://gitter.im/typesafehub/config so that's an available forum to hash things out when you get to it.

**Comment by cprice404** [comment]
We're starting to poke at this a bit now.  A question: I noticed that you have a PR to update your minimum Java version to 8.  Will you have a branch available that doesn't include that change?  Will it be possible for us to target these changes at that branch?  I'm not sure that we are in a spot where we can force our users to bump up higher than 7 just yet.

**Comment by havocp** [comment]
the immediate need was Duration, which I guess is in 7. so would it work to go to 7 only?

**Comment by havocp** [comment]
nevermind, the duration in 7 is some xml thing.

**Comment by cprice404** [comment]
We already have 7 as our minimum version, but 8 would probably be prohibitive for us for another year or so.

**Comment by cprice404** [comment]
Java 8 questions aside:

We've been whiteboarding on the config formatting changes for a little while now, and have some thoughts / questions.

I think the biggest hangup we're having is that it seems like in your description of the situation, you're proposing that the library maintain a big string of all the original text, and then, coupled alongside that string in some new data structure, an array(?) containing information about all of the token offsets in the string?

Assuming we are understanding that properly, and assuming we built such a data structure, then presumably when we wanted to write the Config object back out to a file, we'd do so by traversing that array.  If that's the case, then, when we need to add new nodes into the Config object programatically, we'll have to do a lot of very complex stuff to shift things around in that array... it seems like it might be complex and expensive.  (Maybe we're just misinterpreting your suggestion?)

What we've been talking through seems slightly different from how we're reading your approach.  We'd still retain the String/Stream representation of the original config file text, but we'd attach the start/end offset information to the Tokens or ConfigOrigins or ConfigValue objects directly.  Then, you can manipulate the Config objects just like you normally would; insert new nodes, modify nodes, etc.  When you're done, and you want to write the file back out, you'd start at the root node, write out the opening tokens for it, iterate over its children and write them out, and then write out the closing tokens for the root node.

This would imply a few things:

* Would need to change the implementation of SimpleConfigObject to use a sorted map internally, so that we can preserve the order of the keys.
* Track slightly different offset information for leaf nodes vs. for arrays/maps:
** For leaf nodes, we'd just need to add start offset and end offset (or start token and end token) integer fields to the ConfigValue subclass or ConfigOrigin object
** For arrays / maps, we'd need two pairs of integers: a start/end offset/token for everything leading up to the opening '[' or '{', and a start/end offset/token for everything after the last child and including the closing ']' or '}'.  This would allow us to write out an array value by writing all of the leading tokens, and then iterating over the children to write them out, and then writing out the closing tokens.

Not sure how well I'm explaining this, let me know if that makes sense.  Basically, we're having a hard time imagining what you envisioned the data structure to look like on the "AST/template" side, and how we'd insert new values into it.  And we're wondering if there's some reason why you were thinking we should try to walk that data structure when saving the file out, as opposed to walking over the ConfigValue objects from their root node.

**Comment by havocp** [comment]
Well, to a certain extent this will require playing around and figuring it out - it's too large a task to predict entirely without trying anything. So my macro recommendation would be to mess around a little bit.

Here's why I _think_ but don't know that trying to work with ConfigObject directly will cause problems (tl;dr separation of concerns):

 * of the methods on Config/ConfigObject, virtually none of them are relevant to this scenario, so tangling the code for this concern and the code for ConfigObject's current concerns seems likely to make a large confusing file. I don't know what code from ConfigObject you'd actually need in the load/edit/save scenario.
 * ConfigObject doesn't have to come from a file at all, and can come from files with several different syntaxes
 * it seems problematic to me to mix withFallback() and resolve() into the load/edit/save scenario. those seem like very thorny problems to say the least. Objects need not contain values which were in the original file; they can also contain values from other files and non-files.
 * from a process standpoint, if you're building a large patch over a long time, if it's a separate "pipeline stage" then you have fewer merge conflicts and something easier to review.
 * I'm not 100% sure your proposed approach of walking the config tree writing out the tokens that went with each ConfigValue will work... we would have to avoid including comments, braces, commas, newlines, etc. in the tokens for that, and then where will the comments, braces, commas, and newlines we want to preserve be stored?
 * ConfigObject currently explodes if you try to do much with unresolved nodes, but for an AST document API there's nothing wrong with unresolved nodes

Here's how I was thinking load/edit/save would work. What I'm thinking is that ConfigOrigin (and thus a ConfigValue) points back into the AST data structure, which is specific to a single file. So, to get a new doc to save, you might do something like this: `String saveme = myValue.origin().document().replaceNodeWithValue(myValue.origin().node(),anotherValue).render();` 

That's oversimplified a little, one reason is that `origin().node()` probably doesn't have to exist if the value didn't come from a JSON/HOCON file. So there are various ways the API could handle that.

One advantage of this is that it's very clear that you are going to end up saving a modified version of the file `myValue` came from - the parent document of `myValue`'s document node. While a ConfigObject can be from multiple files, a document node cannot.

Expanding on my above example a little, load/edit/save could look like:
```
 ConfigDocument doc = parse(myfile);
 // edit foo.bar
 Config config = ConfigValueFactory.fromDocument(doc);
 ConfigDocumentNode node = config.getValue("foo.bar").origin().node();
 ConfigDocumentNode newDoc = doc.replaceNodeWithValue(node, value);
 save(file, newDoc.render());
```

(note that all values are immutable, as in the rest of the config API)

You may have inferred that I was talking about a complicated data structure for the document-with-nodes thing; it may turn out complicated, I'm not sure, but I was thinking it's just a tree of nodes and each one has a substring of the doc in it. Unlike the tree in ConfigObject, some of the nodes are purely "syntactic," like comments, braces, and whitespace.

To replace a value you would replace the substring it came from with a render() of the new value. It does not have to be especially efficient, I don't think, especially to start.

So this is a DOM-like API. Unlike the `Config` and `ConfigObject` APIs, it would have no way to do type conversions (like `getInt`, `getMilliseconds`, etc.), it would have no way to do `withFallback`, and no way to `resolve`. It would contain syntactic (braces, whitespace) nodes. For `ConfigValue` instances that originated from one of these JSON/HOCON documents, you could do `value.origin().node()` or something to get back to the original document and potentially modify it.

`ConfigOrigin` for nodes that arise from one of these documents would not necessarily store much data directly, only pointers back into the document for the value and its comments and so on.

I don't know. I think the hard part, for whatever approach, will be changing the parser and the tokenizer around so they preserve more information, while keeping equivalent test coverage. But I could be wrong.

**Comment by cprice404** [comment]
Thanks again for the feedback.  We did start playing around with it yesterday but we're still learning some areas of the code, so we basically only got far enough to start to understand where we'd be able to create the new 'Document' object and where we'd store the reference to it.  It looked like that would probably happen in `Parseable`, around the line where it calls `reader()`:

https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Parseable.java#L193

And we were also thinking that here, we'd add a method to ConfigOrigin called `withSourceText` or `withDocument` or something, which would allow us to attach the new data structure to the ConfigOrigin.  So I think we are with you up to that point.

From there, I think that maybe part of our oversight was the whole 'multi-file' situation.  We don't really have any use-cases for round-tripping the files that would necessarily need to involve more than one file... and I think maybe this is partially because we don't use `include`s, so we weren't thinking about that.  But now I understand why that would potentially be problematic.

So I guess at this point my main confusion is still around what the data structure for this `ConfigDocument` class might look like.  You mention that it'd be "a tree of nodes", but I'm not yet quite understanding what levels/hierarchy would exist.  If the individual nodes are not necessarily representing a config value (because some nodes represent comments, whitespace, etc.) then in what situations to we create sibling nodes vs. child nodes?

I guess maybe we should be trying to think about this like an HTML parser--DOM-like, as you said.  But in our case it's only `[` and `{` that cause new levels in the DOM hierarchy?  Everything else is just siblings?

**Comment by havocp** [comment]
I'm not sure about tree vs. just a list - maybe it doesn't have to be a tree, and it's enough if the derived ConfigObject is a tree. It could be just a glorified version of the current `Iterator<Token>`, where the sequence of tokens preserves _all_ characters found in the original file, and where there are also some methods to replace a subsequence of tokens, and a `value.origin()` would reference the subsequence of tokens that it's derived from. In that world maybe the approach would be to try to change Tokenizer.java so that it doesn't lose any information, and change `Iterator<Token>` into `ConfigDocument` with some useful methods on it such as subsequence replacement. Then the parser would take this new fancier document instead of `Iterator<Token>` here: https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Parser.java#L29

I would not expect you need to modify Parseable.java very much since it's mostly concerned with where the raw stream to be parsed comes from; I would expect you're mostly messing with Parser.java, Tokenizer.java. But who knows.

You could try altering Tokenizer.tokenize to keep all characters found in the original file: https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Tokenizer.java#L51
This would mean adjusting the possible tokens (which would become something more like the "nodes" I was mentioning): https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Tokens.java

Here's an example of where we discard whitespace: https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Tokenizer.java#L87 any character-discarding in Tokenizer.java would need to be fixed.

If you look at the problem as simply making the tokenizer lose zero information, then maybe it's more tractable. We could start with that goal (to be able to read in a file as tokens, then write the tokens back, and get the same file). All private API at first. And then we could tackle the problem of how to associate these tokens with ConfigOrigin, and what the public API should be.

**Comment by cprice404** [comment]
The reason we were looking at `Parseable` is because that seems to be the one place we're guaranteed to have access to the Reader object before any bytes are read from it; and if we needed to maintain a String or other copy of the original document so that we can retain pointers back into it, it seems like it has to happen before any bytes are read by the tokenizer.

If we're talking about switching gears now and looking to keep the bytes in the tokens themselves rather than a new, central `Document` object, then I think you're right that we wouldn't need to mess with `Parseable`.

I'm also all for trying to come up with some incremental approach that allows intermediate PRs/tests along the way, so the Token suggestion sounds appealing.  I do think I'd like to come up with at least some rough dream of what the write API would end up looking like, though, because I think we were toying with this path a bit yesterday and something was throwing us off.

So...
* Tokens retain all original characters they were created from
* AbstractConfigValue.parse is modified to keep a reference to the relevant tokens in some way, probably via ConfigOrigin, 
* then maybe we're back to something like you suggested before?:

```
 ConfigDocumentNode node = config.getValue("foo.bar").origin().node();
 ConfigDocumentNode newDoc = doc.replaceNodeWithValue(node, value);
```

Where 'node()' is basically a sequence of tokens under the covers?  So we're just getting the bytes back from the Token objects directly, no need to keep a separate full copy of the contents of the file like we'd originally been discussing?

One last random question: it seems like the mutators you're hinting at here would be specific to these new "Document"/node objects.  So the API for changing the data in a file would be entirely separate from the API for changing the value of an entry in an in-memory Config representation.  Correct?  The implication being that as a user of the API, you need to know ahead of time whether you're manipulating config for use elsewhere in your program, or for the specific case of modifying the file, and call the appropriate API accordingly?  (Just asking for the sake of clarification, not necessarily raising any opposition :)

Thanks again for taking the time to discuss.

**Comment by havocp** [comment]
I wasn't trying to suggest an alternative to the ConfigDocument idea, instead suggesting that ConfigDocument could be a replacement for that `Iterator<Token>`, and perhaps it could be an evolution of the token list toward something that also supports splicing in new tokens, rendering back out a list of tokens, and can be referenced from an origin(). So as you say, the ConfigDocumentNode is a subsequence of the document's tokens.

I think the concatenation of all tokens should result in a full copy of the original file; that's the invariant that allows load/save cycles without messing up the file formatting. In fact we should probably have a test that specifically tests this for all kinds of weirdly-formatted files.

On the last question, yes, I think the API for modifying a file should be distinct from the API for modifying a Config. The reason is that I don't think the existing API contract of Config can actually be mapped to file modification reliably, and when it can be mapped, it could be inefficient to keep updating the ConfigDocument when you have no intention to save. If we do a tailored API for file modification, then we can make it very clear that you need to pick which file you are modifying (for example) and can't modify a Config that's been merged (`withFallback`) or resolved (`resolve`).

A typical load/edit/save use-case it seems to me would be about a specific file, not about the combined final Config with system properties merged in and everything.

**Comment by cprice404** [comment]
OK.  Thanks for your patience :)

Re: the two separate APIs: ack.  Just making sure I understood.

Re: concatenation of all tokens resulting in full copy of the original file: I agree with that as well.  I think my last point of confusion for the moment is where the bytes will be stored in the object model.  In the earlier parts of our conversation I was thinking that we were talking about taking a copy/snapshot of the full stream/string before we started tokenizing it, and then modifying the Token class to contain start/end offset pointers that would be used to index into that full copy/snapshot.  The way I'm starting to read this now, though, is that maybe the Tokens themselves would just hold on to a copy of the relevant original characters that were read from the Reader during tokenizing.  If that's what you are thinking as well, then I think we could start hacking on that right away.  Does that sound right?

**Comment by havocp** [comment]
it sounds right to me. or at least worth trying :-)

I was just thinking an idea for the first sprint would be to 1) add a render method to go from a list of tokens to a string 2) take the existing tokenizer tests (which just test which tokens we get for a string) and add a check that we render the original string when we render the list of  tokens. this will of course make the tests fail. 3) start extending the parser and Tokens.java until the tests pass.

once we have lossless tokens, we could separately deal with public API and ConfigDocument and ConfigOrigin.

Just trying to break it down into steps we can review and merge so we don't have one giant patch that takes forever.

**Comment by cprice404** [comment]
Awesome, that sounds tractable.  Thanks.

**Comment by havocp** [comment]
this tokenizerTest method is the one I'm thinking of: https://github.com/typesafehub/config/blob/master/config/src/test/scala/com/typesafe/config/impl/TokenizerTest.scala#L14 

it could also do `assertEquals(s, render(expected))` to prove we don't lose anything on round trip for any of the tokenizer tests.

I think that'd be plenty for a first PR, since it'd get us to a place where the parser loses no formatting data.

**Comment by fpringvaldsen** [comment]
@havocp I've been working on adding the ability to reconstruct the source text from a list of Tokens with a render method, and have gotten most of the Tokenizer tests successfully asserting that rendering a TokenIterator exactly produces the original string.

However, I've run into a problem. The code is currently throwing out whitespace between certain values. To achieve lossless tokens, I've had to ensure that whitespace is no longer thrown out, which means that there are now Unquoted Text tokens for all whitespace rather than just for whitespace between simple values. This is causing most of the other tests to fail, as the Parser doesn't know how to handle these new tokens.

I can think of two ways around this. I could either modify the Parser to handle the new whitespace tokens, or I can modify the Tokens themselves so that this whitespace is added to the original text of the succeeding token, rather than existing as its own token.

I'm thinking that modifying the Parser is the correct approach (since other whitespace is stored as Unquoted Text tokens, it seems like whitespace between non-simple values should be stored as Unquoted Text tokens as well). @havocp Is this your preferred approach?

**Comment by havocp** [comment]
You are closer to it than me so just take this as a possibility, but what I might try is to create a new token type like "ignored whitespace", and then the parser skips those tokens. Since the tokenizer can already tell which whitespace is ignored, it may as well go ahead and distinguish, allowing the parser to have very simple logic ("ignore the ignored whitespace"). The token renderer, of course, would still pay attention to these ignored whitespace tokens.

**Comment by cprice404** [comment]
@havocp would you prefer that we move the discussion over to #178 ?

**Comment by havocp** [comment]
no, it doesn't make any difference. we may as well keep it in one place :-)

**Comment by fpringvaldsen** [comment]
@havocp Sounds good! I've implemented a new `IgnoredWhitespace` token class, which the parser ignores.

**Comment by havocp** [comment]
This can be closed along with #272 I guess. Thanks for all the help!
